File: internal/handler/dto/web_search_provider.go
package dto

import (
	"context"
	"time"

	"github.com/Tencent/WeKnora/internal/types"
)

// WebSearchProviderResponse mirrors types.WebSearchProviderEntity for
// response bodies, with the APIKey field removed by construction. Credential
// presence is exposed via the /credentials subresource.
type WebSearchProviderResponse struct {
	ID          string                         `json:"id"`
	TenantID    uint64                         `json:"tenant_id"`
	Name        string                         `json:"name"`
	Provider    types.WebSearchProviderType    `json:"provider"`
	Description string                         `json:"description"`
	Parameters  WebSearchProviderParametersDTO `json:"parameters"`
	IsDefault   bool                           `json:"is_default"`
	CreatedAt   time.Time                      `json:"created_at"`
	UpdatedAt   time.Time                      `json:"updated_at"`
	// Per-field "configured?" map. See MCPServiceResponse.Credentials.
	Credentials map[string]CredentialFieldMetadata `json:"credentials,omitempty"`
}

// WebSearchProviderParametersDTO holds every parameter field except APIKey.
// EngineID, BaseURL, ProxyURL and ExtraConfig are not secrets (they describe
// where to send the request, not how to authenticate) and remain visible.
type WebSearchProviderParametersDTO struct {
	EngineID    string            `json:"engine_id,omitempty"`
	BaseURL     string            `json:"base_url,omitempty"`
	ProxyURL    string            `json:"proxy_url,omitempty"`
	ExtraConfig map[string]string `json:"extra_config,omitempty"`
}

// NewWebSearchProviderResponse converts a stored entity into its response shape.
func NewWebSearchProviderResponse(ctx context.Context, e *types.WebSearchProviderEntity) *WebSearchProviderResponse {
	if e == nil {
		return nil
	}
	params := WebSearchProviderParametersDTO{
		EngineID:    e.Parameters.EngineID,
		BaseURL:     e.Parameters.BaseURL,
		ProxyURL:    e.Parameters.ProxyURL,
		ExtraConfig: e.Parameters.ExtraConfig,
	}
	if !CanViewIntegrationSecrets(ctx) {
		params.ProxyURL = ""
		params.ExtraConfig = nil
	}
	return &WebSearchProviderResponse{
		ID:          e.ID,
		TenantID:    e.TenantID,
		Name:        e.Name,
		Provider:    e.Provider,
		Description: e.Description,
		Parameters:  params,
		IsDefault:   e.IsDefault,
		CreatedAt:   e.CreatedAt,
		UpdatedAt:   e.UpdatedAt,
		Credentials: map[string]CredentialFieldMetadata{
			"api_key": {Configured: e.Parameters.APIKey != ""},
		},
	}
}

func NewWebSearchProviderResponses(ctx context.Context, es []*types.WebSearchProviderEntity) []*WebSearchProviderResponse {
	out := make([]*WebSearchProviderResponse, 0, len(es))
	for _, e := range es {
		out = append(out, NewWebSearchProviderResponse(ctx, e))
	}
	return out
}

// WebSearchProviderAssignmentDTO mirrors
// types.WebSearchProviderAssignmentInfo — the admin-console view of one
// "this workspace uses this service" row.
type WebSearchProviderAssignmentDTO struct {
	TenantID   uint64    `json:"tenant_id"`
	TenantName string    `json:"tenant_name"`
	IsDefault  bool      `json:"is_default"`
	AssignedAt time.Time `json:"assigned_at"`
}

// SystemWebSearchProviderResponse is the platform-catalog shape served by
// /system/admin/web-search-providers: the base provider plus its workspace
// assignments. IsDefault is deliberately absent — a default is a
// per-workspace property carried by the assignment rows, not by the
// platform row.
type SystemWebSearchProviderResponse struct {
	ID          string                         `json:"id"`
	TenantID    uint64                         `json:"tenant_id"`
	Name        string                         `json:"name"`
	Provider    types.WebSearchProviderType    `json:"provider"`
	Description string                         `json:"description"`
	Parameters  WebSearchProviderParametersDTO `json:"parameters"`
	CreatedAt   time.Time                      `json:"created_at"`
	UpdatedAt   time.Time                      `json:"updated_at"`
	// Per-field "configured?" map. See MCPServiceResponse.Credentials.
	Credentials  map[string]CredentialFieldMetadata `json:"credentials,omitempty"`
	Assignments  []WebSearchProviderAssignmentDTO   `json:"assignments"`
}

// NewSystemWebSearchProviderResponse converts a platform entity plus its
// assignment rows into the admin-console response shape.
func NewSystemWebSearchProviderResponse(
	ctx context.Context,
	e *types.WebSearchProviderEntity,
	assignments []types.WebSearchProviderAssignmentInfo,
) *SystemWebSearchProviderResponse {
	if e == nil {
		return nil
	}
	params := WebSearchProviderParametersDTO{
		EngineID:    e.Parameters.EngineID,
		BaseURL:     e.Parameters.BaseURL,
		ProxyURL:    e.Parameters.ProxyURL,
		ExtraConfig: e.Parameters.ExtraConfig,
	}
	if !CanViewIntegrationSecrets(ctx) {
		params.ProxyURL = ""
		params.ExtraConfig = nil
	}
	dtos := make([]WebSearchProviderAssignmentDTO, 0, len(assignments))
	for _, a := range assignments {
		dtos = append(dtos, WebSearchProviderAssignmentDTO{
			TenantID:   a.TenantID,
			TenantName: a.TenantName,
			IsDefault:  a.IsDefault,
			AssignedAt: a.AssignedAt,
		})
	}
	return &SystemWebSearchProviderResponse{
		ID:          e.ID,
		TenantID:    e.TenantID,
		Name:        e.Name,
		Provider:    e.Provider,
		Description: e.Description,
		Parameters:  params,
		CreatedAt:   e.CreatedAt,
		UpdatedAt:   e.UpdatedAt,
		Credentials: map[string]CredentialFieldMetadata{
			"api_key": {Configured: e.Parameters.APIKey != ""},
		},
		Assignments: dtos,
	}
}

File: frontend/src/api/web-search-provider.ts
import { get, post, put, del } from '@/utils/request'

// WebSearchProviderEntity represents a configured web search provider instance
export interface WebSearchProviderEntity {
  id?: string
  tenant_id?: number
  name: string
  provider: 'bing' | 'google' | 'duckduckgo' | 'tavily' | 'ollama' | 'baidu' | 'searxng' | 'keenable' | 'zhipu' | 'metaso' | 'exa' | 'serpbase'
  description?: string
  parameters: {
    // api_key is never returned by the server in this shape; it lives behind
    // the /credentials subresource. Kept on the type so the initial create
    // POST can still include it.
    api_key?: string
    engine_id?: string
    base_url?: string
    proxy_url?: string
    extra_config?: Record<string, string>
  }
  is_default?: boolean
  // Per-field configured? metadata from the main response.
  credentials?: Record<WebSearchCredentialField, { configured: boolean }>
  created_at?: string
  updated_at?: string
}

// Workspace assignment of a platform search service (000095 platform rework):
// a service is configured once at platform level and shared with workspaces
// through these rows; at most one assigned service per workspace is the
// default (flagged here, else the earliest assignment acts as one).
export interface WebSearchProviderTenantAssignment {
  tenant_id: number
  tenant_name: string
  provider_id?: string
  is_default: boolean
  assigned_at?: string
}

// Platform catalog row as returned by the /system/admin endpoints — the
// tenant-scoped entity shape plus the service's workspace assignments.
// is_default is deliberately absent: defaults live on assignments only.
export interface SystemWebSearchProviderEntity extends WebSearchProviderEntity {
  assignments?: WebSearchProviderTenantAssignment[]
}

// WebSearchProviderTypeInfo describes metadata for a provider type
export interface WebSearchProviderTypeInfo {
  id: string
  name: string
  requires_api_key: boolean
  // Keyless-by-default providers that still accept an optional key (e.g. Keenable).
  supports_optional_api_key?: boolean
  requires_engine_id?: boolean
  requires_base_url?: boolean
  supports_proxy?: boolean
  description?: string
  docs_url?: string
  config_fields?: WebSearchProviderConfigField[]
}

export interface WebSearchProviderConfigField {
  key: string
  label: string
  label_key?: string
  type: 'select'
  required?: boolean
  default?: string
  description?: string
  description_key?: string
  options?: Array<{ label: string; label_key?: string; value: string }>
}

// ---- 管理面（000095 平台化）：搜索服务由系统管理员在平台目录配置一次，
// 再按空间分配。全部走 /system/admin 镜像路由，仅系统管理员可用。 ----

// Create a new platform web search provider
export function createWebSearchProvider(data: Partial<SystemWebSearchProviderEntity>) {
  return post('/api/v1/system/admin/web-search-providers', data)
}

// List the whole platform search-service catalog (admin console)
export function listSystemWebSearchProviders(): Promise<SystemWebSearchProviderEntity[]> {
  return get('/api/v1/system/admin/web-search-providers').then((res: any) => {
    if (res.success && Array.isArray(res.data)) {
      return res.data
    }
    return []
  })
}

// Get a single platform web search provider by ID (with assignments)
export function getWebSearchProvider(id: string) {
  return get(`/api/v1/system/admin/web-search-providers/${id}`)
}

// Update an existing platform web search provider
export function updateWebSearchProvider(id: string, data: Partial<SystemWebSearchProviderEntity>) {
  return put(`/api/v1/system/admin/web-search-providers/${id}`, data)
}

// Delete a platform web search provider (its assignments are purged server-side)
export function deleteWebSearchProvider(id: string) {
  return del(`/api/v1/system/admin/web-search-providers/${id}`)
}

// Get available provider types (for dynamic form rendering) — console copy
export function listWebSearchProviderTypes(): Promise<WebSearchProviderTypeInfo[]> {
  return get('/api/v1/system/admin/web-search-providers/types').then((res: any) => {
    if (res.success && res.data) {
      return res.data
    }
    return []
  })
}

// ---- 空间读路径（保持原端点）：聊天开关 / 智能体编辑器解析当前空间的
// 已分配服务列表（is_default 为空间视角的有效默认）。 ----

// List the web search providers assigned to the current workspace
export function listWebSearchProviders() {
  return get('/api/v1/web-search-providers')
}

// ----------------------------------------------------------------------------
// Web search provider credential subresource (platform catalog).
// ----------------------------------------------------------------------------

export type WebSearchCredentialField = 'api_key'

export interface WebSearchCredentialsResponse {
  fields: Record<WebSearchCredentialField, { configured: boolean }>
}

export async function putWebSearchProviderCredentials(
  id: string,
  body: Partial<Record<WebSearchCredentialField, string>>,
): Promise<WebSearchCredentialsResponse> {
  const response: any = await put(`/api/v1/system/admin/web-search-providers/${id}/credentials`, body)
  return (response.data ?? response) as WebSearchCredentialsResponse
}

export async function deleteWebSearchProviderCredentialField(
  id: string,
  field: WebSearchCredentialField,
): Promise<void> {
  await del(`/api/v1/system/admin/web-search-providers/${id}/credentials/${field}`)
}

// Test a web search provider connection.
// If id is provided, tests the existing saved provider.
// If data is provided, tests with raw credentials (no persistence).
export function testWebSearchProvider(id?: string, data?: { provider: string; parameters: any }): Promise<any> {
  if (id) {
    return post(`/api/v1/system/admin/web-search-providers/${id}/test`, {})
  }
  return post('/api/v1/system/admin/web-search-providers/test', data || {})
}

// ----------------------------------------------------------------------------
// Workspace assignments of a platform search service.
// ----------------------------------------------------------------------------

export function listWebSearchProviderTenantAssignments(id: string): Promise<WebSearchProviderTenantAssignment[]> {
  return get(`/api/v1/system/admin/web-search-providers/${id}/tenant-assignments`).then((res: any) => {
    if (res.success && Array.isArray(res.data)) {
      return res.data
    }
    return []
  })
}

// Replace the full assignment list of a service. rows carries the target
// state; a workspace missing from the list is unassigned.
export function updateWebSearchProviderTenantAssignments(
  id: string,
  rows: Array<{ tenant_id: number; is_default: boolean }>,
) {
  return put(`/api/v1/system/admin/web-search-providers/${id}/tenant-assignments`, { assignments: rows })
}
